File: src/rdruid_analyzer/web/cache.py
import json
from pathlib import Path


class ResultCache:
    def __init__(self, cache_dir: Path = Path("data/results_cache")):
        self._dir = cache_dir
        self._dir.mkdir(parents=True, exist_ok=True)

    def _key_path(self, code: str, fight_id: int, player: str) -> Path:
        safe_player = player.lower().replace(" ", "_")
        return self._dir / f"{code}_{fight_id}_{safe_player}.json"

    def get(self, code: str, fight_id: int, player: str) -> dict | None:
        path = self._key_path(code, fight_id, player)
        if not path.exists():
            return None
        return json.loads(path.read_text())

    def set(self, code: str, fight_id: int, player: str, data: dict) -> None:
        path = self._key_path(code, fight_id, player)
        path.write_text(json.dumps(data))

File: go-backend/internal/web/cache.go
package web

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type ResultCache struct {
	dir string
}

func NewResultCache(dir string) *ResultCache {
	os.MkdirAll(dir, 0755)
	return &ResultCache{dir: dir}
}

func (c *ResultCache) keyPath(code string, fightID int, player string) string {
	safePlayer := strings.ToLower(strings.ReplaceAll(player, " ", "_"))
	return filepath.Join(c.dir, fmt.Sprintf("%s_%d_%s.json", code, fightID, safePlayer))
}

func (c *ResultCache) Get(code string, fightID int, player string) map[string]any {
	path := c.keyPath(code, fightID, player)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var result map[string]any
	if err := json.Unmarshal(data, &result); err != nil {
		return nil
	}
	return result
}

func (c *ResultCache) Set(code string, fightID int, player string, data map[string]any) {
	path := c.keyPath(code, fightID, player)
	bytes, _ := json.Marshal(data)
	os.WriteFile(path, bytes, 0644)
}
